geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java

```
@@ -2251,48 +2237,54 @@ void checkQueueSizeConstraint() throws InterruptedException {
-                      this.maxQueueSizeHitCount = 0;
-                    }
-                    ++this.maxQueueSizeHitCount;
-                    this.region.checkReadiness(); // fix for bug 37581
-                    // TODO: wait called while holding two locks
-                    this.permitMon.wait(CacheClientNotifier.eventEnqueueWaitTime);
-                    this.region.checkReadiness(); // fix for bug 37581
-                    // Fix for #51400. Allow the queue to grow beyond its
-                    // capacity/maxQueueSize, if it is taking a long time to
-                    // drain the queue, either due to a slower client or the
-                    // deadlock scenario mentioned in the ticket.
-                    reconcilePutPermits();
-                    if ((this.maxQueueSizeHitCount % logFrequency) == 1) {
-                      logger.info("Resuming with processing puts ...");
-                    }
-                  } catch (InterruptedException ex) {
-                    // TODO: The line below is meaningless. Comment it out later
-                    this.permitMon.notifyAll();
-                    throw ex;
-                  }
-                }
-              }
-            } // synchronized (this.permitMon)
-          } // if (putPermits <= 0)
-          --putPermits;
-        } // synchronized (this.putGuard)
+        long startTime = System.currentTimeMillis();
+        if (putPermits.get() <= 0) {
+          synchronized (this.permitMon) {
+            long duration = (System.currentTimeMillis() - startTime);
+            checkQueueSizeConstraintCore(duration);
+          } // synchronized (this.permitMon)
+        } // if (putPermits <= 0)
+        putPermits.decrementAndGet();
+      }
+    }
+
+    /* Do not call this method directly from anywhere except checkQueueSizeConstraint */
+    private void checkQueueSizeConstraintCore(long duration) throws InterruptedException {
+      if (reconcilePutPermits() <= 0) {
```

since (in my design) the caller has already decd putPermits, then we can stop waiting in this method if reconcilePutPermits() >= 0. So I think you want this comparison to be < 0.

---

If I use < 0 this code will fail every single time. It prefers <= 0 so to fail early it would be if (reconcilePutPermits() > 0) { 
return
}
My observations show reconcilePutPermits returns 0, -1, or -2.

---

Done.